accounts/README.md

```diff
@@ -1,10 +1,123 @@
+| address      | to        | The address of the contract to be called                                                                                                                                                                                                                                                                                                                                                  |
+| uint256      | value     | The amount of Native Token to be sent                                                                                                                                                                                                                                                                                                                                                     |
+| bytes memory | data      | The call data                                                                                                                                                                                                                                                                                                                                                                             |
+| uint8        | operation | The operation type to be executed. `0` for CALL, `1` for DELEGATECALL. The DELEGATECALL operation can only be executed by a Plugin with root access permissions. More information about transaction types can be found in the [Manager contract](https://github.com/safe-global/safe-core-protocol-specs/blob/2bffd759dd12be5583594f302d97c35e0ab9fcf5/manager/README.md) specifications. |
+
+The account MUST execute a corresponding `operation` to the `to` address with the provided `value` and `data` parameters. The account MUST return a tuple of `(bool success, bytes memory returnData)`. 
+It is RECOMMENDED that the account supports the DELEGATECALL operation.
+
+#### Hooks
+
+Hooks do not require any specific interface to be implemented by the account. To use hooks, use the interface defined in the
+[Safe{Core} Protocol Manager specification](../manager/README.md)
+
+#### Function Handlers
+
+The account MUST implement a `fallback` function, which forwards all the calldata
+to the Safe{Core} Protocol Manager AND appends the 20 bytes of the `msg.sender` address to the calldata.
+
+An example implementation of such a `fallback` function is:
+```Solidity
+    fallback() external {
+        bytes32 slot = SAFE_CORE_PROTOCOL_MANAGER_SLOT;
+        assembly {
+            function allocate(length) -> pos {
+                pos := mload(0x40)
+                mstore(0x40, add(pos, length))
+            }
```

Comment:
Since we are guaranteed to return in this `assembly` block, and that the memory usage is ephemeral to this block, I don't think writing to reading or writing to 0x40 is needed. You can probably:

```
calldatacopy(0, 0, calldatasize())
mstore(calldatasize(), shl(96, caller()))
let success := call(gas(), protocol, 0, 0, add(calldatasize(), 20), 0, 0)
returndatacopy(0, 0, returndatasize())
if iszero(success) {
    revert(0, returndatasize())
}
return(0, returndatasize())
```

For large calldata and returnsdata, this could have a significant gas impact as the memory cost function is quadratic in the `msize()` of the call, which would be here `~= calldatasize() + returndatasize()` and with my proposal `~= max(calldatasize(), returndatasize())`.

Reply:
Not important for the specification per se - so not blocking an approval, just a potential improvement to the reference implementation.

Reply:
I agree that your implementation is more optimized. However, if you compile using the optimizer, it may lead to problems because the compiler may use the memory to move some stack variables and expects the memory to be in  "a well-defined state"; see here: https://docs.soliditylang.org/en/v0.8.20/assembly.html#memory-safety

IIRC, the original Safe implementation was similar to yours, but then we had to change it because the contracts wouldn't compile with 0.8.x and the optimizer enabled (and you had to enable the optimizer to fit in the bytecode size limit). More details about the problem here https://github.com/safe-global/safe-contracts/issues/544

Reply:
> However, if you compile using the optimizer, it may lead to problems

I think this is explicitly not true for `assembly` blocks. In particular, from the documentation you linked (emphasis my own):

> While we recommend to always respect Solidity’s memory model, **inline assembly allows you to use memory in an incompatible way**. Therefore, moving stack variables to memory and additional memory optimizations are, by default, globally disabled in the presence of any inline assembly block that contains a memory operation or assigns to Solidity variables in memory.

If we want to allow the optimiser to touch this code, then we should mark it with:

```solidity
assembly ("memory-safe") {
    // ...
}
```

Reply:
Ahh, now I remember that this particular block didn't compile with solidity optimizer enabled (it'd just error with stack too deep). I directly referenced that snippet in the conversation with the solidity team: https://github.com/safe-global/safe-contracts/issues/544#issuecomment-1528768835

I'm okay with changing it to the unsafe version - with explicitly mentioning its unsafety 😄

Reply:
Also it seems like one unsafe block affects all the other blocks, quote from https://github.com/safe-global/safe-contracts/issues/544#issuecomment-1498952802:
> Yes. All blocks that are included in the code a contract is built from. So e.g. if you have two different contracts in a file and they don't inherit from each other, blocks in one do not influence the other. But if both use a library function imported from a different file and that function has an unsafe block, both are affected.

Reply:
Ah, you are right, it **globally** disables these optimisations (not just in the function). In that case, we should keep your version (otherwise you can severely limit the compiler), but in that case, can we annotate the assembly block with `assembly ("memory-safe")` since it is with this version?

Reply:
I'll add the comment annotation that's also compatible with earlier solidity versions:
```
/// @solidity memory-safe-assembly
```